### PR TITLE
Linkserver: adding west runner and board support

### PR DESCRIPTION
**Pull request by yvanderv**
This PR adds a west runner for NXP's LinkServer (flash and debug) as well as support for selected boards.

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
Hi @yvanderv, please do not close your old pull request and open a new one every time you make a change. Instead, use 'git push --force' to update your existing pull request.

**Comment by yvanderv** [comment]
> 

I usually don't.  The PR was closed due to my mistake...

File: doc/develop/flash_debug/host-tools.rst [hunk 1]
```diff
@@ -200,6 +200,43 @@ Windows PATH. A specific bossac executable can be used by passing the
 
    WSL is not currently supported.
 
+LinkServer Debug  Host Tools
+****************************
+
+Linkserver is a utility for launching and managing GDB servers for NXP debug probes, which also provides a command-line target flash programming capabilities. Linkserver can be used with NXP MCUXpresso for Visual Studio Code implementation, with custom debug configurations based on GNU tools or as part of a headless solution for continuous integration and test. Linkserver can be used with MCU-Link, LPC-Link2, LPC11U35-based and OpenSDA based standalone or on-board debug probes from NXP. The Linkserver installer also includes the firmware update utilities for MCU-Link and the LPCScrypt utility for use with LPC-Link2. Linkserver can also be installed using the MCUXpresso Installer.
```
Comment (mbolivar-nordic) [inline doc/develop/flash_debug/host-tools.rst:206]: Zephyr documentation style is to fill the paragraph by adding newlines, rather than having very long lines in a text file. (This applies to the entire diff in this PR)

File: doc/develop/flash_debug/host-tools.rst [hunk 2]
```diff
@@ -200,6 +200,43 @@ Windows PATH. A specific bossac executable can be used by passing the
...
+
+LinkServer is compatible with the following debug probe:
```
Comment (mbolivar-nordic) [inline doc/develop/flash_debug/host-tools.rst:208]: probes

File: doc/develop/flash_debug/host-tools.rst [hunk 3]
```diff
@@ -200,6 +200,43 @@ Windows PATH. A specific bossac executable can be used by passing the
...
+Notes:
+
+
+1. Probes can be listed with LinkServer:
+
+.. code-block:: console
+
+   LinkServer probes
+
+2. Use the LinkServer west runner   '--probe' option to pass the probe index.
```
Comment (mbolivar-nordic) [inline doc/develop/flash_debug/host-tools.rst:228]: Please use double backticks so sphinx doesn't render `--` as an em dash:

```
``probe``
```

File: doc/develop/flash_debug/probes.rst [hunk 4]
```diff
@@ -58,6 +58,33 @@ onboard debug probe may have limitations, such as lack of support for advanced
 debuggers or high-speed tracing. You may need to adjust jumpers to prevent the
 onboard debug probe from interfering with the external debug probe.
 
+.. _lpclink2-cmsis-dap-onboard-debug-probe:
+
+LPC-LINK2 CMSIS DAP Onbpard Debug Probe
```
Comment (mbolivar-nordic) [inline doc/develop/flash_debug/probes.rst:63]: onboard

File: scripts/west_commands/runners/linkserver.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
+# Copyright 2023 NXP
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+# Based on jlink.py
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:5]: Propagate copyright headers from this file

File: scripts/west_commands/runners/linkserver.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+DEFAULT_LINKSERVER_EXE = 'Linkserver.exe' if sys.platform == 'win32' else 'LinkServer'
+DEFAULT_LINKSERVER_GDB_PORT =  3333
+DEFAULT_LINKSERVER_SEMIHOST_PORT = 3334
+
+#class ToggleAction(argparse.Action):
+#
+#    def __call__(self, parser, args, ignored, option):
+#        setattr(args, self.dest, not option.startswith('--no-'))
+
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:27]: Remove

File: scripts/west_commands/runners/linkserver.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    def capabilities(cls):
+        return RunnerCaps(commands={'flash', 'debug', 'debugserver', 'attach'},
+                          dev_id=True, flash_addr=True, erase=True,
+                          tool_opt=True, file=True)
+
+    #@classmethod
+    #def dev_id_help(cls) -> str:
+    #    return '''Device identifier. Use it to select the probe
+    #              of the device connected over USB.'''
+
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:76]: Remove. In general, do not submit commented-out or debug code.

File: scripts/west_commands/runners/linkserver.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+                                 linkserver=args.linkserver,
+                                 dt_flash=args.dt_flash,
+                                 erase=args.erase,
+                                 probe=args.probe,
+                                 semihost_port=args.semihost_port,
+                                 gdb_port=args.gdb_port,
+                                 override=args.override,
+                                 tui=args.tui, tool_opt=args.tool_opt)
+
+## UPDATE
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:114]: What's going on here? Is there more work to do?

File: scripts/west_commands/runners/linkserver.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            ls_output=self.check_output(linkserver_version_cmd)
+            self._linkserver_version = str(ls_output.split()[1].decode())
+
+        return self._linkserver_version
+
+    @property
+    def linkserver_version_str(self):
+
+        return self.linkserver_version
+
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:139]: This seems unnecessary; why is it here?

File: scripts/west_commands/runners/linkserver.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+    @property
+    def supports_thread_info(self):
+        # TODO: check if linkserver needs this??
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:143]: Please resolve this

File: scripts/west_commands/runners/linkserver.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+        return True
+
+    def do_run(self, command, **kwargs):
+
+        # Convert commander to a real absolute path. We need this to
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:148]: Are you sure this is necessary? In my experience, shutil.which() (which underlies self.require()) returns absolute paths. Did you have a different experience?

File: scripts/west_commands/runners/linkserver.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+        # be able to find the shared library that tells us what
+        # version of the tools we're using.
+        self.linkserver = os.fspath(
+            Path(self.require(self.linkserver)).resolve())
+        self.logger.info(f'LinkServer Version: {self.linkserver_version_str}')
+
+        #rtos = self.thread_info_enabled and self.supports_thread_info
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:155]: remove debug code

File: scripts/west_commands/runners/linkserver.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+        if command == 'flash':
+            self.flash(**kwargs)
+        else:
+            linkserver_cmd = ([self.linkserver] +
+                              ["gdbserver"]    +
+                              ["--probe", "#"+str(self.probe) ] +
+                              ["--gdb-port", str(self.gdb_port )] +
+                              ["--semihost-port", str(self.semihost_port) ] +
+			      self.override_cli +
+                              [self.device])
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:166]: Indentation looks wrong here

File: scripts/west_commands/runners/linkserver.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+            if command in ('debug', 'attach'):
+            # self.print_gdbserver_message()
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:169]: debug code

File: scripts/west_commands/runners/linkserver.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+                if self.elf_name is  None or not os.path.isfile(self.elf_name):
+                    raise ValueError('Cannot debug; elf file required')
+
+                gdb_cmd = ([self.gdb_cmd] +
+                           self.tui_arg +
+                          [self.elf_name] +[
+                          '-ex', 'target remote {}:{}'.format(self.gdb_host, self.gdb_port)])
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:176]: Indentation

File: scripts/west_commands/runners/linkserver.py [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+
+                self.run_server_and_client(linkserver_cmd, gdb_cmd)
+
+            elif command == 'debugserver':
+                if self.gdb_host:
+                    raise ValueError('Cannot run debugserver with --gdb-host')
+
+                self.check_call(linkserver_cmd)
+
+            return
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:192]: remove unnecessary line

File: scripts/west_commands/runners/linkserver.py [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+        self.logger.debug("flash erase command = " + str(linkserver_cmd))
+        self.check_call(linkserver_cmd)
+
+    def _build_override_cli(self):
+
+        override_cli = []
+
+        if self.override is not None:
+            for ov in self.override:
+                override_cli = (override_cli + [ "-o", str(ov)])
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:207]: remove whitespace after `[`

File: scripts/west_commands/runners/linkserver.py [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+
+        if self.erase:
+            self.do_erase()
+
+        if self.bin_name is not None and os.path.isfile(self.bin_name):
+            if self.dt_flash:
+                load_addr = self.flash_address_from_build_conf(self.build_conf)
+            else:
+                self.logger.critical("no load flash address could be found...")
+                raise RuntimeError("missing argument")
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/linkserver.py:223]: What's going on here? It seems like you always want to flash using partition info from devicetree/.config. The error doesn't seem related to a missing argument.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Comments apply to the commit adding a runner only

File: scripts/west_commands/runners/linkserver.py (hunk 18 above)
Comment (yvanderv) [reply scripts/west_commands/runners/linkserver.py:223]: I believe all of your comments have been addressed.

File: boards/common/linkserver.board.cmake [hunk 19]
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (mmahadevan108) [inline boards/common/linkserver.board.cmake:1]: missing NXP Copyright

**Comment by mmahadevan108** [comment]
@yvanderv, please update the board documentation files to add information about Linkserver

**Comment by dleach02** [comment]
@mbolivar-nordic, will you be able to return to this PR to verify your change requests?

File: scripts/west_commands/runners/linkserver.py (hunk 13 above)
Comment (yvanderv) [reply scripts/west_commands/runners/linkserver.py:166]: Checked on my end, it looks good.

**Review by mmahadevan108 — DISMISSED** [review]

**Comment by mmahadevan108** [comment]
@mbolivar-nordic , can you help re-review.

**Comment by dleach02** [comment]
> @mbolivar-nordic , can you help re-review.

Marti contacted me and said he will get on this next week when back from vacation.

**Review by mbolivar-nordic — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@yvanderv can you rebase? (merge conflict)

**Comment by yvanderv** [comment]
> @yvanderv can you rebase? (merge conflict)

@fabiobaltieri, all done.

**Review by mmahadevan108 — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]
